### PR TITLE
refactor(bucket): Move 'Find-Manifest' and 'list_buckets' to 'buckets'

### PR DESCRIPTION
**Pull request by niheaven**
#### Description

`find_manifest()` and `Find-Manifest()` (former `locate()`) have similar functionality and now be merged into `Find-Manifest()`.

Also move `list_buckets()` from `libexec\scoop-bucket.ps1` to `lib\buckets.ps1`.

#### Motivation and Context

Discussed with @rashil2000 

#### How Has This Been Tested?

`scoop bucket` and `scoop 

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

**Comment by rashil2000** [comment]
Nice change!

Since we're refactoring bucket methods, can we also incorporate https://github.com/ScoopInstaller/Scoop/pull/3348? This will allow us to close #2123

**Comment by niheaven** [comment]
Okay, wait a moment.

**Comment by niheaven** [comment]
Done. Now check for duplicated buckets.

File: lib/buckets.ps1 [hunk 1]
```diff
@@ -103,6 +103,24 @@ function Find-Manifest($app, $bucket) {
...
 }
 
+function Convert-RepositoryUri {
+    [CmdletBinding()]
+    param (
+        [Parameter(Mandatory, Position = 0, ValueFromPipeline = $true)]
+        [String] $Uri
+    )
+
+    process {
```
Comment (rashil2000) [inline lib/buckets.ps1:113]: We can omit the `process` block for such a small function

Comment (niheaven) [reply lib/buckets.ps1:113]: Enabling pipeline needs explicit `process`, and in fact, if not specified, the script block is put under `begin`.

Comment (rashil2000) [reply lib/buckets.ps1:113]: Oh, didn't know that 😅

**Comment by rashil2000** [comment]
image

The bucket-add function should stop at the first error (`ERROR github.com/ScoopInstaller/Extras is not a valid Git URL!`), no?

**Comment by niheaven** [comment]
So you mean we should assume default protocol? The error you met is no protocol provided (it is http? https? ftp? ftps? ssh?)

**Comment by rashil2000** [comment]
I meant that we're showing the error twice. When we get an error (any error) from the `Convert-RepositoryUri` function, we should stop the `add_bucket` function (instead of proceeding to the `Checking repo...` step)

**Comment by niheaven** [comment]
Oh, sorry, haven't noticed that, will change it later.

File: lib/buckets.ps1 [hunk 2]
```diff
@@ -117,6 +117,8 @@ function Convert-RepositoryUri {
             $Matches.provider, $Matches.user, $Matches.repo -join '/'
         } else {
             error "$Uri is not a valid Git URL!"
+            error "Please see https://git-scm.com/docs/git-clone#_git_urls for valid ones."
+            exit 1
```
Comment (rashil2000) [inline lib/buckets.ps1:121]: Exiting from script might not be desired if we use this function somewhere else. Instead we can check something like last error code or return value of function.

**Comment by rashil2000** [comment]
Since the location of `Find-Manifest` has changed, has the import of this function in all other source files been taken care of?

**Comment by niheaven** [comment]
> Since the location of Find-Manifest has changed, has the import of this function in all other source files been taken care of?

`buckets.ps1` is imported in every `lib-exec` scripts, so it's not a problem.

**Comment by lorengordon** [comment]
Traced this patch as the cause of this issue #4917

**Pull request merged**